EIPS/eip-7303.md

```diff
@@ -0,0 +1,132 @@
+
+1. Smart contracts implementing the [ERC-7303](./eip-7303.md) standard MUST represent the privilege required by the role as an ERC-721 token. The tokens that represent privileges are called `control tokens` in this ERC. The `control token` can be any type of token, and its transactions may be recursively controlled by another `control token`.
+2. To associate the required `control token` with the role, the address of the previously deployed contract for the `control token` MUST be used.
+3. To ascertain whether an account possesses the necessary role, it SHOULD be confirmed that the balance of the `control token` exceeds 0, utilizing the `balanceOf` method defined in ERC-721.
+4. To grant a role to an account, a `control token` representing the privilege SHOULD be minted to the account using `safeMint` method defined in [ERC-5679](./eip-5679.md).
+5. To revoke a role from an account, the `control token` representing the privilege SHOULD be burned using the `burn` method defined in ERC-5679.
+6. A role in a compliant smart contract is represented in the format of `bytes32`. It's RECOMMENDED the value of such role is computed as a `keccak256` hash of a string of the role name, in this format: `bytes32 role = keccak256("<role_name>")`. such as `bytes32 role = keccak256("MINTER")`.
+  
+## Rationale
+
+TBD
+
+## Backwards Compatibility
+
+This ERC is designed to be compatible for [ERC-721](./eip-721), [ERC-1155](./eip-1155), and [ERC-5679](./eip-5679) respectively.
+
+## Reference Implementation
+
+The following presents a straightforward example of a contract that implements a modifier. This modifier checks if an account possesses the necessary role, and the contract includes a function that grants a specific role to a designated account.
+  
+```solidity
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+
+contract TokenController {
+    mapping (bytes32 => address []) private _controlTokens;
+
+    modifier onlyHasToken(bytes32 r, address u) {
+        require(_checkHasToken(r,u), "TokenController: not has a required token");
+        _;
+    }
+
+    /**
+     * @notice Grant a role to user who owns a control token specified by the contract ID. 
+     * Multiple calls are allowed, in this case the user must own at least one of the 
+     * specified token.
+     * @param r byte32 The role which you want to grant.
+     * @param c address The address of contract ID of which token the user required to own.
+     */
+    function _grantRoleByToken (bytes32 r, address c) internal {
+        _controlTokens[r].push(c);
+    }
+
+    function _checkHasToken (bytes32 r, address u) internal view returns (bool) {
+        for (uint i = 0; i < _controlTokens[r].length; i++) {
```

Comment:
It may be more gas efficient to store the array of addresses from `_controlTokens[r]` in memory of the functions scope as opposed to accessing state on each iteration

Reply:
Yes! Thank you for your valuable comments! I've incorporated it into the revised version.